### DISCUSSION TITLE
Implementing the TRAPI Inference Engine

### DISCUSSION THREAD

**Discussion by vemonet** [NCATSTranslator/ReasonerAPI#238] [Q&A]
One of the new feature will require KPs to handle descendants of BioLink concepts, so we will need to also build an Inference Engine. Which needs to be well integrated to the TRAPI Query Engine.

For this we could use the BL lookup API from the SRI:

https://bl-lookup-sri.renci.org/apidocs/#/lookup/get_bl__concept__descendants

For example, to get descendants of `biolink:ChemicalSubstance` : https://bl-lookup-sri.renci.org/bl/chemical_substance/descendants?version=latest

Unfortunately the BioLink lookup API does not use standard BioLink identifiers: we need to convert `biolink:ChemicalSubstance` to `chemical_substance` which require to add more hardcoded tweaks in a system which should rely on standards

Also we need to send 1 API query everytime we need to get info on the BioLink model

**Comment by edeutsch** [comment]
Hi Vincent, two comments. First:

> Unfortunately the BioLink lookup API does not use standard BioLink identifiers: we need to convert biolink:ChemicalSubstance to chemical_substance which require to add more hardcoded tweaks in a system which should rely on standards

It is my understanding that biolink:ChemicalSubstance *IS* the standard. This is the format required in TRAPI 1.0 and beyond. Just "chemical_substance" was used back in TRAPI 0.9 until it was discovered that this was not the Biolink standard. If you go here:
https://biolink.github.io/biolink-model/docs/ChemicalSubstance.html
you'll that the URI for this concept is biolink:ChemicalSubstance 

Second,
we have some code here that uses the API that you mention and uses caching so that only the first lookup is slow, then fast after that. This could/should be cleaned up to be more elegant, but you're welcome to use this or steal ideas/code from it. I was going to make it fancier and probably faster storage mechanism than requests_cache (store a pickle of the hierarchy and load on first instance), but that is not yet implemented. It also includes some code to handle conflation of different types (e.g. biolink:Drug and biolink:ChemicalSubstance are effectively not usefully distinguished for us and therefore explicitly conflated)
https://github.com/RTXteam/RTX/blob/master/code/ARAX/NodeSynonymizer/category_manager.py
